[macOS] Fix input[type=search] rendering in vertical writing mode

pxlcoder opened the pull request:
#### 64d1b1b4705a751713e9738ef37d43c7f7731368
<pre>
[macOS] Fix input[type=search] rendering in vertical writing mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=248334">https://bugs.webkit.org/show_bug.cgi?id=248334</a>
<a href="https://rdar.apple.com/102658361">rdar://102658361</a>

Reviewed by Tim Nguyen.

Native search fields are not height resizable, and have shadows,
preventing their use in vertical writing mode. Instead, fall back
to drawing native text fields, which are height resizable.

Additionally, ensure that the cancel and results buttons are displayed
correctly in vertical writing mode.

* Source/WebCore/html/shadow/TextControlInnerElements.cpp:

253691@main introduced logic to hide the cancel and results buttons when
authors specify `appearance: textfield` on search inputs. Preserve this
functionality, but also allow cases where the user agent applies
`appearance: textfield` for rendering purposes (in this case, in vertical
writing mode), by comparing `appearance` in addition to `effectiveAppearance`.

(WebCore::searchFieldStyleHasExplicitlySpecifiedTextFieldAppearance):
(WebCore::SearchFieldResultsButtonElement::resolveCustomStyle):
(WebCore::SearchFieldCancelButtonElement::resolveCustomStyle):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::adjustStyle):

Set an effective appearance of `TextField` for search fields using a vertical
writing mode.

* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::searchFieldShouldAppearAsTextField const):
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::searchFieldShouldAppearAsTextField const):
(WebCore::RenderThemeMac::adjustSearchFieldDecorationPartStyle const):

While the results and cancel buttons are not being rotated, ensure there is not
an excessive amount of padding at the logical top of the control in cases where
the results button is hidden.

Canonical link: <a href="https://commits.webkit.org/270415@main">https://commits.webkit.org/270415@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/29986487502b8fb1b0bfa1ac64110e6d9e00dbd1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25416 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4020 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26701 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27530 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23313 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25692 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5748 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1457 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23482 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25661 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21933 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28108 "Built successfully") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22877 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28987 "Passed tests") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23240 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26806 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2617 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/869 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3987 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3066 "Built successfully") | | [⏳ 🛠 jsc-mips ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3249 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2956 "Built successfully") | | [⏳ 🧪 jsc-mips-tests ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/shadow/TextControlInnerElements.cpp:
```diff
@@ -254,13 +254,13 @@ std::optional<Style::ResolvedStyle> SearchFieldResultsButtonElement::resolveCust
         return elementStyle;
     }
-    if (appearance != StyleAppearance::SearchField) {
+    if (effectiveAppearance != StyleAppearance::SearchField && effectiveAppearance != StyleAppearance::TextField) {
```
These checks are not super pretty nor self explanatory IMO.

We should probably have a static method in this file that takes a RenderStyle and says whether the effectiveAppearance is default based on the writing mode and the effectiveAppearance.

pxlcoder replied:
I'll add a static method with a good name for clarity. Though the method won't consider `writing-mode`, since that's a macOS specific behavior. It's more about the original appearance the author specified.

nt1m commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -204,6 +204,11 @@ void RenderTheme::adjustStyle(RenderStyle& style, const Element* element, const
         style.setEffectiveAppearance(appearance);
     }
 
+    if (appearance == StyleAppearance::SearchField && searchFieldShouldAppearAsTextField(style)) {
+        appearance = StyleAppearance::TextField;
+        style.setEffectiveAppearance(appearance);
+    }
```
I somewhat wished we had all of this logic in autoAppearanceForElement

pxlcoder replied:
I wanted to keep the special case separate, similar to the `isControlStyled` cases. Going to leave this as-is unless you feel strongly.

nt1m commented on Source/WebCore/rendering/RenderThemeMac.mm:
```diff
@@ -1276,8 +1281,13 @@ static FloatRect inflateRect(const FloatRect& rect, const IntSize& size, const i
     IntSize size = sizeForSystemFont(style, resultsButtonSizes());
-    style.setWidth(Length(size.width() - emptyResultsOffset, LengthType::Fixed));
-    style.setHeight(Length(size.height(), LengthType::Fixed));
+    int widthOffset = 0, heightOffset = 0;
```
WebKit code style wants one declaration per line

pxlcoder replied:
Fixed.

nt1m commented on Source/WebCore/html/shadow/TextControlInnerElements.cpp:
```diff
@@ -254,13 +260,14 @@ std::optional<Style::ResolvedStyle> SearchFieldResultsButtonElement::resolveCust
-    if (appearance != StyleAppearance::SearchField) {
+
+    auto effectiveAppearance = shadowHostStyle->effectiveAppearance();
+    if (effectiveAppearance != StyleAppearance::SearchField && effectiveAppearance != StyleAppearance::TextField) {
```
Would be good to add a comment like:
// By default, input[type=search] can use either the searchfield or textfield appearance depending on the platform and writing mode. Only adjust the style when that default is used.

pxlcoder replied:
Added.

pxlcoder committed:
[macOS] Fix input[type=search] rendering in vertical writing mode
https://bugs.webkit.org/show_bug.cgi?id=248334
rdar://102658361

Reviewed by Tim Nguyen.

Native search fields are not height resizable, and have shadows,
preventing their use in vertical writing mode. Instead, fall back
to drawing native text fields, which are height resizable.

Additionally, ensure that the cancel and results buttons are displayed
correctly in vertical writing mode.

* Source/WebCore/html/shadow/TextControlInnerElements.cpp:

253691@main introduced logic to hide the cancel and results buttons when
authors specify `appearance: textfield` on search inputs. Preserve this
functionality, but also allow cases where the user agent applies
`appearance: textfield` for rendering purposes (in this case, in vertical
writing mode), by comparing `appearance` in addition to `effectiveAppearance`.

(WebCore::searchFieldStyleHasExplicitlySpecifiedTextFieldAppearance):
(WebCore::SearchFieldResultsButtonElement::resolveCustomStyle):
(WebCore::SearchFieldCancelButtonElement::resolveCustomStyle):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::adjustStyle):

Set an effective appearance of `TextField` for search fields using a vertical
writing mode.

* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::searchFieldShouldAppearAsTextField const):
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::searchFieldShouldAppearAsTextField const):
(WebCore::RenderThemeMac::adjustSearchFieldDecorationPartStyle const):

While the results and cancel buttons are not being rotated, ensure there is not
an excessive amount of padding at the logical top of the control in cases where
the results button is hidden.

Canonical link: https://commits.webkit.org/270415@main
